src/ivector/xvector.cc

```diff
@@ -0,0 +1,140 @@
+// ivector/xvector.cc
+
+// Copyright 2016     Daniel Povey
+//                    David Snyder
+
+// See ../../COPYING for clarification regarding multiple authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ivector/xvector.h"
+
+namespace kaldi {
+
+void ComputeXvectorObjfAndDeriv(
+    const CuMatrixBase<BaseFloat> &xvector_pairs,
+    const CuSpMatrix<BaseFloat> &S,
+    BaseFloat b, CuMatrixBase<BaseFloat> *deriv_xvector,
+    CuVector<BaseFloat> *deriv_S_and_b, BaseFloat *tot_objf,
+    BaseFloat *tot_weight) {
+
```

Comment:
OK, here's a plan for more efficient computation.
firstly, in the following I have not thought carefully about variable names.
you can if you want keep the old version for debugging purposes somewhere- e.g. put it in the -test.cc file.
- copy S to a temporary CuMatrix (since we don't have matrix-multiply with symmetric-packed on CUDA).
- compute the product P =  xvector_pairs * S. 
- compute two matrices: Q = xvector_pairs * xvector_pairs',
   and R = xvector_pairs * P'.
Note: if you want you can use SymAddMat2 in computing Q, since the args are identical, and then do CopyLowerToUpper- it might be slightly faster.
- compute T = Q - R + b, which is the matrix of scores (log-odds).

At this point, define a function to compute the objf from T and the derivs w.r.t. the elements of T.  For now let's ignore the fact that it's symmetric and compute things for both the upper and lower parts, and later divide by 2-- anyway this phase likely won't dominate.
Here is my draft of the documentation for this function:
/*  This function is used in computing the objective function and derivatives in 
    xvector training.
   @param [in] scores   'scores' is a symmetric matrix of scores which are to be
              interpreted as log-odds (according to the model) of pairs coming from the
              same class, so scores(i, j) is the model's log p(same/different) for elements i and j
              of the original minibatch of input.
   @param [out] objf_terms   objf_terms is a matrix of the same dimension as 'scores
               whose elements we will sum to get the objective function for this minibatch.  
              We assume that the data in 'scores' has been arranged in such
              a way that indexes (0, 1) are from the same class; (2, 3) are from the same class and
              so on, but the classes are otherwise unique.   This function computes the appropriate
              contributions to the objective function, as follows. 
                   if i == j:   objf_terms(i, j)== 0       # the same exact element is not scored
                   elsif i%2 == j%2:  objf_terms(i, j) = log(p(same)) = -log(1 + exp(-scores(i, j))
                   else:    objf_terms(i, j) = 1/scores.NumRows() * log(p(different))
                                                      = -1/scores.NumRows() * log(1+exp(scores(i,j))
    @param [out] objf_derivs    Element (i,j) of this matrix is the derivative of objf_terms(i,j)
                                         with respect to scores(i, j).  */
  void ComputeXvectorObjfFromScores(const CuMatrixBase<BaseFloat> &scores,
                                                                CuMatrixBase<BaseFloat> *objf_terms,
                                                                CuMatrixBase<BaseFloat> *objf_derivs);
 Note: this function will have its own CUDA kernel.  Pegah can help with this if you need.
 You can actually declare this function in cu-math.h and define it in cuda-math.cc.  
 Copy and modify the code for the Sigmoid kernel- but you'll have some if-statements
 and you'll need to add new function arguments.  Actually maybe changing the DiffSigmoid
 kernel and making one of the inputs an output would be easier, since that also will have
 3 matrix arguments.   Inside the .cc file, after the #endif, instead of backing off to
 another named function you can just put the non-CUDA version of the code right
 there in the function.  The files you'll have to edit for this are cu-math.{h,cc},
 cu-kernels-ansi.h, cu-kernels.h and (most crucially) cu-kernels.cu.  It will be 
 mostly copy-and-modify; search for diff_sigmoid.

----
OK, now let's return to ComputeXvectorObjfAndDeriv.  After you have computed the matrix
of objf_terms you can just sum it to compute the actual objective function (let 'tot_weight'
be S.NumRows() and fix the documentation).  And as for objf_derivs, you can use that
to start backpropagating through the bits of the computation that you just did, to compute
derivatives w.r.t xvector_pairs and S and b.  Again you'll have to be careful about the symmetry in
S; the function ScaleDiag in class PackedMatrix may come in useful here.

Your current code can all go in the -test.cc file.